File: design_patterns/creational/abstract-factory/abstract-factory.py
from __future__ import annotations
from abc import ABC, abstractmethod

class WidgetFactory(ABC):
    @abstractmethod
    def create_scrollbar(self) -> ScrollBar:
        pass

    @abstractmethod
    def create_window(self) -> Window:
        pass

class MotiWidgetFactory(WidgetFactory):
    def create_scrollbar(self) -> MotiScrollBar:
        return MotiScrollBar()

    def create_window(self) -> MotiWindow:
        return MotiWindow()

class PMWidgetFactory(WidgetFactory):
    def create_scrollbar(self) -> PMScrollBar:
        return PMScrollBar()

    def create_window(self) -> PMWindow:
        return PMWindow()

class Window(ABC):
    @abstractmethod
    def whoami(self) -> str:
        pass

class MotiWindow(Window):
    def whoami(self) -> str:
        return "MotiWindow"

class PMWindow(Window):
    def whoami(self) -> str:
        return "PMWindow"

class ScrollBar(ABC):
    @abstractmethod
    def whoami(self) -> str:
        pass

class MotiScrollBar(ScrollBar):
    def whoami(self) -> str:
        return "MotiScrollBar"

class PMScrollBar(ScrollBar):
    def whoami(self) -> str:
        return "PMScrollBar"


# PLAYING WITH EXAMPLE

def client_code(factory: WidgetFactory) -> None:
    """
    Client code works only with factories and products through abstract types.
    This lets you pass any factory or product subclass to the client code
    without breaking it.
    """
    window = factory.create_window()
    scrollbar = factory.create_scrollbar()
    
    print(f"{window.whoami()}")
    print(f"{scrollbar.whoami()}", end="\n")

if __name__ == "__main__":
    """
    Client code can work with any concrete factory class
    """
    print("Client: testing client code with MotiWidgetFactory type:")
    client_code(MotiWidgetFactory())

    print("Client: testing client code with PMWidgetFactory type:")
    client_code(PMWidgetFactory())

File: container2/design_patterns/creational/abstract-factory/abstract-factory.js
class WidgetFactory {
    // Abstract class
    createScrollBar = function() {}
    createWindow = function() {}
}

class MotiWidgetFactory extends WidgetFactory {
    createScrollBar = function() {
        return new MotiScrollBar()
    }
    createWindow = function() {
        return new MotiWindow()
    }
}

class PMWidgetFactory extends WidgetFactory {
    createScrollBar = function() {
        return new PMScrollBar()
    }
    createWindow = function() {
        return new PMWindow()
    }
}

class Window {
    // Abstract Class
    whoami() { console.log('I am a Windows') }
}

class MotiWindow extends Window {
    whoami() { console.log('I am a Moti Window') }
}

class PMWindow extends Window {
    whoami() { console.log('I am a PM Window') }
}

class ScrollBar {
    // Abstract Class
    whoami() { console.log('I am a ScrollBar') }
}

class MotiScrollBar extends ScrollBar {
    whoami() { console.log('I am a Moti ScrollBar') }
}

class PMScrollBar extends ScrollBar {
    whoami() { console.log('I am a PM ScrollBar') }
}

/**
 * Example code:
 * Client code works only with products through abstract types, created by factories.
 * This pattern lets you pass any factory or product subclass to the client code, at runtime, without breaking it.
 */

function clientCode(factory) {
    let myWindow = factory.createWindow()
    let myScrollBar = factory.createScrollBar()

    myWindow.whoami()
    myScrollBar.whoami()
}

clientCode(new MotiWidgetFactory())
clientCode(new PMWidgetFactory())
